velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,217 @@
+          } else if (it->second) {
+            it->second = false;
+            rawNewIndices[indicesCursor] = i;
+            // [Sort] if it is the first non-null element, do nothing;
+            // otherwise, swap with the previous non-null element if the current
+            // is less
+            if (!((indicesCursor == rawOffsets[row] + 1 && numOfNulls > 1) ||
+                  indicesCursor == rawOffsets[row]) &&
+                (value < prevDuplicateValue)) {
+              std::swap(
+                  rawNewIndices[indicesCursor],
+                  rawNewIndices[indicesCursor - 1]);
+            }
+
+            indicesCursor++;
+            prevDuplicateValue = value;
+          }
+        }
+      }
+
+      uniqueMap.clear();
+      rawSizes[row] = indicesCursor - rawOffsets[row];
+    });
+
+    newIndices->setSize(indicesCursor * sizeof(vector_size_t));
+    auto newElements =
+        BaseVector::transpose(newIndices, std::move(elementsVector));
+
+    // Prepare and return result set.
+    auto resultArray = std::make_shared<ArrayVector>(
+        pool,
+        caller->type(),
+        nullptr,
+        rowCount,
+        std::move(newOffsets),
+        std::move(newLengths),
+        std::move(newElements),
+        0);
+
+    context->moveOrCopyResult(resultArray, rows, result);
+  }
+
+ private:
+  // Validate element type inside the vector
+  static void validateElementType(const TypeKind& kind) {
+    VELOX_USER_CHECK(
+        kind == TypeKind::TINYINT || kind == TypeKind::SMALLINT ||
+            kind == TypeKind::INTEGER || kind == TypeKind::BIGINT ||
+            kind == TypeKind::VARCHAR,
+        "array_duplicates only takes argument of type array(varchar) or array(bigint)");
+  }
+};
+
+// Validate number of parameters and types.
+void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_USER_CHECK_EQ(
+      inputArgs.size(), 1, "array_duplicates requires exactly one parameter");
+
+  auto arrayType = inputArgs.front().type;
+  VELOX_USER_CHECK_EQ(
+      arrayType->kind(),
+      TypeKind::ARRAY,
+      "array_duplicates requires arguments of type ARRAY");
+}
+
+// Create function template based on type.
+template <TypeKind kind>
+std::shared_ptr<exec::VectorFunction> createTyped(
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 1);
+
+  using T = typename TypeTraits<kind>::NativeType;
+  return std::make_shared<ArrayDuplicatesFunction<T>>();
+}
+
+// Create function.
+std::shared_ptr<exec::VectorFunction> create(
+    const std::string& /* name */,
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  validateType(inputArgs);
+  auto elementType = inputArgs.front().type->childAt(0);
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      createTyped, elementType->kind(), inputArgs);
+}
+
+// Define function signature.
+std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+  // array(T) -> array(T)
```

Comment:
I just noticed that this signature is too broad. It says that the function accepts any array(T), but this function allows only a subset of T, e.g. tinyint, smallint, integer, bigint and varchar. Please, update.